library/modules/EventManager.cpp

```diff
@@ -527,12 +539,33 @@ static void manageJobCompletedEvent(color_ostream& out) {
     int32_t tick0 = eventLastTick[EventType::JOB_COMPLETED];
     int32_t tick1 = df::global::world->frame_counter;
 
-    multimap<Plugin*,EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
-    unordered_map<int32_t, df::job*> nowJobs;
-    for ( df::job_list_link* link = &df::global::world->jobs.list; link != nullptr; link = link->next ) {
-        if ( link->item == nullptr )
-            continue;
-        nowJobs.emplace(link->item->id, link->item);
+    multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
+    std::vector<JobCompleteData> nowJobs;
+    // predict the size in advance, this will prevent or reduce rehashing elements.
+    nowJobs.reserve(prevJobs.size());
+    for (const auto jobPtr : df::global::world->jobs.list) {
+        auto& job = *jobPtr;
+        // Can we pack these into a single integer?
+        nowJobs.emplace_back(job.id, job.completion_timer, (bool)job.flags.bits.repeat);
```

Comment:
Maybe add a constructor that takes the job as argument? In fact, I think you only one constructor either way.

Reply:
Either way, this isn't really a priority. There are bigger fish to fry in this PR.